### ISSUE TITLE
Session-Scoped Fixtures are not Session-Scoped with Pytest-Xdist

### ISSUE THREAD

**Issue by elijahbal** [pytest-dev/pytest-xdist#271]
I am fairly new to this project. I very recently migrated a software project test-suite from nosetest to pytest, mainly because of the Xdist benefits I had heard of.

The problem is that my Tests are depending on a big fixture setup (table creation + heavy loading of data) that I would like to share across *all* my tests.

The current Xdist behaviour is as follow:
1. Collect test
2. Split tests amongst the user-defined number of processes
3. Launch a pytest session for each worker

Obviously, if each test depend on a heavy fixture, then multiplying the number of creation per number of worker is not going to help.

Additionally, it simply breaks the expected pytest behaviour for 'session' scoped fixtures.

I think this should be fairly simple to address this problem although I didn't take a really deep look into it. If you need help to solve this problem, I am more than willing to contribute if you feel this suggestion for improvement is relevant.

Greetings.

**Comment by nicoddemus** [comment]
Hi @elijahbal,

The problem is that each worker lives in a different process, so it would be hard to share a single session scoped fixture between the workers in a general manner.

For one you have the problem of serializing any object returned by a fixture so it can be sent to another process, on demand. `pickle` might seem like the obvious solution to this problem, but a lot of objects are not pickable.

Another problem, much harder IMHO, is how do you keep the object returned by the fixture synchronized among the worker processes? After all a session fixture might return a mutable object, so its state must be synchronized somehow as well.

Another problem is that one must also deal with parallelism, given that tests are now running in parallel in separate processes. The resources probably will need to be synchronized using a lock.

Those are the reasons from the top of my head why it would be hard to implement the proper scope rules in a way that session scoped fixtures are only instantiated once per-session, as opposed to once per-worker like it is today.

**Comment by elijahbal** [comment]
Happy to have so fast an answer. Yes, the problem of having a different space for each process to live on is a tough one, and parallelizing the computation comes at the cost of a greater complexiy for handling the process, as far I understand.

First of all, I would like to address each of the problem you mention, because answering each of them can be of a great help to solve the overall problem. 

First of all, concerning the ability to serialize and to large objects between processes, I agree with you, in my experience, it is hard to serialize complex objects with pickle.

Of the second problem, keeping a synchronized state between worker, I will tell more thereafter.

And for the third problem, yes, indeed, the more direct-and-obvious approach to have global session object is to use a system lock at the operating system level, regardless of the programming language.

For your point 1 and 2, I want to adress some part of the problem and I think it would not appear a so big difficulty. The use case of spawning large fixture object is usually not to share big collections of references (aka very big objects), but rather to prepare the underlying system to be in a specific case. We could include for example:

- spawning many containerized application for a large micro-services software
- setting up a database in a pristine state with a clean table structure
- Gathering a big set of data on the local storage drive from a collection of network connections
- etc.

So here the problem is not so much to share a big object between process but to ensure that the fixture preparation process is run only once from a single process. The synchronization between processes of such objects is not really important, because basically these are objects that are precisely meant to be accessed in a concurrent fashion (webserver, database, etc).

From this point of view, I think (but maybe I am wrong) that the xdist initialization step (collection of test and the preparation of the test sessions with distribution of tests amongst runner) is especially well suited to perform such a task.

So maybe a good compromise would be to have a special xdist marker for such fixtures that would care about the initialization process. As for the return value of such fixtures, it would be probably best to enforce for such special fixtures that they return at first call an immutable primitive object (eg a string or at max a tuple of strings) that would be kept in memory by xdist.

Does it seem to you a good idea ? And does it seem to you something doable in practice given the current xdist codebase ?

**Comment by jrodriguezntt** [comment]
I'm having a similar issue with session scoped fixtures when n > 1. My project does some of the things @elijahbal points out above.

The teardown section in the session fixture is being executed more than once, and worst of it, out of order: before other tests have ended.

I'm looking for a way to signal "there are no more pending tests" to workaround this. If such case, a worker can check if it's the last one and effectively execute such section or do nothing. Is there a way? Would be very helpful!

**Comment by elijahbal** [comment]
No idea actually how this could be done. The quickest workaround I could come with is to use a separate resource for each one of the xdist process.

**Comment by nicoddemus** [comment]
Hi folks, sorry for the late response, this issue slipped through the cracks.

@elijahbal 

> Does it seem to you a good idea ? And does it seem to you something doable in practice given the current xdist codebase ?

Not really, and I don't think just because of pytest-xdist but how pytest itself is structured, the session scope fixture instantiated in some worker will be destroyed once that worker runs out of tests, regardless if there are other workers still needing that resource for example.

@elijahbal and @jrodriguezntt I can't find a way to do what you need with fixtures, but perhaps using plain hooks executed only on `master` can be responsible for creating the shared resource and destroying it (`pytest_sessionstart/finish` come to mind)? This has the problem of course that you might end up initializing a resource that won't even be used (for example if `-k` is used to select some tests), but might be worth giving this a try.

**Comment by jrodriguezntt** [comment]
I've managed to workaround this by doing nothing in the teadown section (after yield). A cleanup bash script is executed after py.test, but it's a pity, because this (IMHO) breaks the logic of the `scope='session'` option (once per session). I understand the technical problem, nonetheless.
One approach could be to use semaphores or any other shared resource if workers are running in the same machine. Another one is to use a shared resource like memchached, but this might be a bit too complicated (not sure if it's worth it).
The hook on master is also a good idea, provided it's the last one to execute (once the workers has finished). But I have no experience using them. Any suggestions will be welcome! :)

**Comment by RonnyPfannschmidt** [comment]
each process is a own session, while its a nice to have feature to have a scope that spanns processes, its also important to note that there is no good sane and simple way to implement it - the proposed approaches are all no-gos on techical merit

**Comment by mherbert** [comment]
While the following workarounds are probably obvious to people discussing here, they may not be obvious to xdist noobs like me who googled and landed here while skipping the [documentation](https://pypi.org/project/pytest-xdist/). So mentioning them briefly:

```
    if ("PYTEST_XDIST_WORKER" not in os.environ or
            os.environ["PYTEST_XDIST_WORKER"] == "gw0"):
         <initialize something>
         <create some empty file>
    else:
          <wait for empty file>
```
```
    if ("PYTEST_XDIST_WORKER_COUNT" not in os.environ or
            os.environ["PYTEST_XDIST_WORKER_COUNT"] == 1):
         ....
```

> Another problem, much harder IMHO, is how do you keep the object returned by the fixture synchronized among the worker processes? After all a session fixture might return a mutable object, so its state must be synchronized somehow as well.

In my case and I guess many others, the part of the fixture that is time-consuming to download is read-only. Modern languages encourage read-only structures for obvious... concurrency reasons https://doc.rust-lang.org/book/second-edition/ch03-01-variables-and-mutability.html

Note in my case the "better" fix would probably be to move the downloads outside Python and to "make test" and "make clean" targets - but that is more work.

**Comment by RonnyPfannschmidt** [comment]
how about simply putting a http download cache into `.pytest-cache`

**Comment by pv** [comment]
How do you synchronize access to `.pytest-cache` (or `request.config.cache.makedir`)?
Does pytest have something builtin for this?

**Comment by rogerdahl** [comment]
I use Postgres, and found that `create database from template` is much faster than creating a database fixture from scratch. So I have each process launched by xdist create its database from template. If the template does not exist, I use a system wide lock to have the xdist process that gains the lock first create the template (from JSON files), while blocking the others from proceeding. When the template is ready, the lock is released, all the other processes find that the template already exists and create their database instances directly from the template. I went from maybe 10 minutes to a couple of seconds getting the database fixtures set up.

```
@pytest.yield_fixture(scope='session')
def db_setup(request):
  db_set_unique_db_name(request)
  with posix_ipc.Semaphore(
      '/{}'.format(__name__), flags=posix_ipc.O_CREAT, initial_value=1
  ):
    if not db_exists():
      db_create_blank()
      db_migrate()
      db_populate_by_json()
      db_migrate()

  db_create_from_template()

  yield

  db_drop()
```

Notes about the lock:

`posix_ipc` is at `https://pypi.org/project/posix_ipc/` (I'm not affiliated with it).

Regular multiprocessing.Lock() context manager did not work for me here. Also tried creating the lock at module scope, and also directly calling acquire() and release(). It's probably related to how the worker processes relate to each other when launched by pytest-xdist as compared to what the multiprocessing module expects.

**Comment by smasterfree** [comment]
or use a file lock? only when you finish your init ENV , others process will go on?

**Comment by eode** [comment]
Don't you already have IPC/an event system between Manager and Workers?

Couldn't you use the existing events system to handle fixtures `xdist_group_lock`, `xdist_global_lock` and `xdist_group` and `xdist_global`, where:

- `xdist_xxx_lock` are locks that can be acquired by workers
- `xdist_xxx` are dicts restricted to hashable, serializable data, where small size is recommended

..?

Then you could easily do:

```python
def some_test(xdist_group_lock, xdist_group):
    xdist_group_lock.acquire()   # blocks
    if not xdist_group['initialized']:
        initialize_things()
        xdist_group['initialized'] = True
    xdist_group_lock.release()
```

..and those primitives aren't *too* hard to do over a network -- particularly one you already have an event system over.  Yes, it would slow things down a tad for some people that use it, particularly if they abuse it, but there's still a net benefit in most cases.

**Comment by nicoddemus** [comment]
@eode it is probably possible. I believe this can even be implemented outside xdist as a proof of concept first.

**Comment by eode** [comment]
I've done this kind of control with other pytest plugins by using `multiprocessing.lock()` set up in `conftest.py`, but those plugins used the multiprocessing module, and apparently loaded `conftest.py` before forking / creating other processes.

However, I don't know the method that xdist uses to distribute work.  However xdist does it, making a lock in `conftest.py` doesn't work.  My suspicion is that xdist either doesn't use the multiprocessing module, or it forks before `conftest.py` is imported.

So, proof-of-concept works -- however, some kind of distributed `Lock` (and probably a distributed `dict`) would be needed, but it needs to be via a means of communication that is just as guaranteed as your normal IPC -- so, using whatever IPC you normally use.

As to the general principle -- it's been tested, and works.

**Comment by RonnyPfannschmidt** [comment]
xdist currently uses execnet and is incompatible with multiprocessing mechanisms #302 is supposed to elevate that

**Comment by nicoddemus** [comment]
> I believe this can even be implemented outside xdist as a proof of concept first.

I should have been clearer, but I meant that providing such a fixture, which uses a file lock behind the scenes, is possible.

**Comment by rogerdahl** [comment]
@eode Network distributed testing should be covered by another type of lock, since implicit locking between machines probably would be unwanted as often as not. The new type of lock should probably use the database. I think testing based on other network shared resources is esoteric enough that it should be handled on a case by case basis by the user.

**Comment by eode** [comment]
Not that a random user matters much, but FYI this is the reason I chose and use a less-developed framework for parallel testing.

I've implemented a lock, and a dict, with execnet before, it's definitely doable -- and if the client isn't able to communicate with the server via channels, then all is pretty much lost there anyways, as far as execnet is concerned.

What do you mean about implicitly locking?  That seems like a bad idea.  I was more thinking of a 'sessionlock', 'modulelock', and 'classlock' fixture that would need to be explicitly called and used by a test.

The raw fact of the matter is that sometimes, synchronization is needed between clients that aren't on the same physical system with access to the same data.  Communication is also sometimes needed.

**Comment by rogerdahl** [comment]
With implicit locking, I was thinking about session scoped fixtures automatically being session scoped across multiple machines on the network.

Support for various more specific scopes, like a network wide session scope, would be nice. Might be better than exposing locks directly?

**Comment by eode** [comment]
Hey, that's not a bad idea.  An 'xdist-session' scope that executes before even distributing tests would solve a lot of use cases that would otherwise require synchronization.  It would even remove the need for locking in many cases.


Speaking of scopes, i discovered in pytest that the session scope executes *after* tests have been imported and organized.  The intuitive place to mock globally seemed to be the session scope, but not so.  If mocking `somelib.bar` from the session scope, you need to track down every case of `from somelib import bar` and patch that individually.    

That makes mocking of some python libraries a pain, since there's no scope that supports post-conftest, pre-code-import.  I think I ended up piggy-backing on the config plugin hook or similar.  That worked, but coming from outside the project, it wasn't easy figuring out what to do, and was very counterintuitive and not clearly documented -- which really stands out in pytest, because pytest is mostly intuitive and well-documented.  :-)

Also to pytest's credit, the fix once found was nice and succinct.

Anyways, to the point - Additional scopes sound like a good idea to me.

**Comment by neXussT** [comment]
I am also having this issue with n>1. My session fixtures are executed multiple times. I am testing with AWS resources, and keep hitting "Resource Conflict", due to multiple fixtures trying to write environmental variables to the same lambda at the same time.
This could be solved, as stated above with the session fixtures running once.

**Comment by RonnyPfannschmidt** [comment]
a key problem with that is that python has  no exchange mechanism that allows to do this safely over multiple processes

i believe something like pytest_configure_resources that sets up the global resources would help as a hook, design work on that one is needed

**Comment by david-fliguer-ws** [comment]
Hi, after a while I am still trying to deal with this

If I understand correctly there is still no "official" solution for the issue of having to perform a series of actions only once before the tests start to run in parallel,

Is still the best workaround is to "split" the run in two, one "fake" run which makes all the preparations and then the actual parallel run ???

**Comment by RonnyPfannschmidt** [comment]
yup, please note that with the current design of pytest-xdist its just the wrong thing to do in many cases

**Comment by rogerdahl** [comment]
@RonnyPfannschmidt Could you elaborate?

**Comment by RonnyPfannschmidt** [comment]
@rogerdahl in situations where xdist is used to distribute tests across networks its very unclear where to set up fixtures

as such setting up fixtures only once is simply a no go, so the fixture life-cycles are bound to the process

pytest doesn't yet have a concept for resources obtained for a specific test run

**Comment by david-fliguer-ws** [comment]
Hi @RonnyPfannschmidt 

I have been working for a while with pytest but I never actually entered to the core of it, but as a regular question: Is not the purpouse of the session scoped fixtures to do a setup that needs to be done once before all the tests ? (Something like beforeAll in JUnit methodology)

In that case, besides the technical aspects/limitations, from a logical point of view, do you think should xdist "support" this ?

Again I really love how xdist is so easy to use but with this limitation I feel kind of stuck (Specially because I already have a project that makes plenty of logic in a session scoped fixture)

Thanks !!!

David

**Comment by RonnyPfannschmidt** [comment]
as far as pytest core is concerned, each worker process is a session, so the session scope is "correct" so to speak

the idea of resources that are obtained "per test run" are slightly different and currently not modelled in pytest itself

**Comment by nicoddemus** [comment]
Would also point to the docs which show how to make a session-scoped fixture which executes only once:

https://github.com/pytest-dev/pytest-xdist#making-session-scoped-fixtures-execute-only-once

The main issue is that, as @RonnyPfannschmidt said, each worker is its own session, and there's no way currently to serialize/unserialize arbirtrary objects and send them over the wire from a worker to the master process. The docs linked above show how you can accomplish it though assuming the object produced by the fixture is read-only.

**Comment by david-fliguer-ws** [comment]
Hi nicoddemus, thanks for the response, I will take a look to that solution.

If I add a similar solution and I perform a pytest run without xdist then nothing should be affected ?

**Comment by nicoddemus** [comment]
> If I add a similar solution and I perform a pytest run without xdist then nothing should be affected ?

Yep! 👍

**Comment by kapilt** [comment]
I'm interested in seeing what can be done with this, at least for setting up a robust pattern to enable folks to self-service. assuming non-distributed ( -n auto) and serializing simple data structures (json) to disk for communication. the file lock example works for resource creation, but with resources that need cleanup, its a bit more interesting. an individual worker that setup a session scope won't be able to initiate a teardown without knowing that all tests that reference it are already executed. I've been looking at serializing a test map of test node ids to fixture names, and then a test execution log on the master, and then on session, thats checked in the session scoped finalizer before initiating teardown. Does that approach sound reasonable or any alternatives come to mind?

**Comment by RonnyPfannschmidt** [comment]
personally i w ould do that kind of setup/teardown in a way that the resource creation happens in pytest_configure and the   fixtures just return objects  created from json in any case

then pytest_unconfigure would do the cleanup

**Comment by kapilt** [comment]
thanks for the suggestion, unfortunately I'm not sure that really works for my use cases, it requires apriori knowledge of all session fixtures, pre collection. my intent was really more of a pytest plugin that works with xdist while ensuring expected per run semantics around non function scopes, afaics I won't have knowledge of the extant fixtures pre-collection, moreover the fixture setup is fairly expensive on time though parallelizable (running terraform to provision infrastructure), so serializing it in the master is a bit wasteful for the goal wrt to xdist usage. Your suggestion brought an additional doc question of missing knowledge wrt to understanding xdist's worker vs manager delegation of hooks / responsibilities. it looks like the master doesn't do any collection, but does bootstrap/initialization, xdist hooks, and relay reporting, the workers do collection, running, and reporting, so building up that map of node ids to fixture names would need to be serialized w/ lock from a worker node.

**Comment by RonnyPfannschmidt** [comment]
i see, that is indeed not nicely doable atm,
a pytest plugin providing a resoucemanager concept, and providing the gateway connections to the master process resource manager may be a interesting side-channel to spearhead that

then a fixture could request a specific resource

ideally the resource system would reuse the fixture system in some way, but thats currently not doable sanely

**Comment by kapilt** [comment]
fwiw, i went ahead and built something a set of fixtures with once per test run locking compatible with xdist (single node), using file locking semantics, and a test log, and 
https://github.com/cloud-custodian/pytest-terraform

 it works on versions !5.4 at the moment, it fails there till there's a release with @RonnyPfannschmidt 's fix for https://github.com/pytest-dev/pytest/pull/6992
as it currently uses the tmpdir base for setting up lock file directories on coordination.

**Comment by kousu** [comment]
> Hey, that's not a bad idea. An 'xdist-session' scope that executes before even distributing tests would solve a lot of use cases that would otherwise require synchronization. It would even remove the need for locking in many cases.

I second/third this. I am working around this in my tests by [`fcntl.lockf()` per @apenwarr's research](https://apenwarr.ca/log/20101213) but pytest-xdist is already managing all the fixtures and workers so it should be able to coordinate them.

When xdist is in the mix, I understand "session" has to split into at least three scopes:

1. globally, per `testrun` (e.g. installing a database schema inside mysql that all the tests can share)
2. per-node (e.g. downloading test data to the machine)
3. per-worker (e.g. generating in-RAM data)

My assumption was `@fixture(scope="session")` meant #1 and I was very confused to figure out it meant #3. I still think that's what it should mean, even though this thread helped me realize that in most of my tests [what I really need is per-node scoping](https://github.com/neuropoly/spinalcordtoolbox/blob/7efbbb2db90afdb577f836958413bb70fd29c00e/conftest.py#L31).

Is it possible to hook pytest in a way to overload the meaning of "session" and add these two other scopes too:

1. `scope="session"`
2. `scope="node"`
3. `scope="worker"`

?

**Comment by kousu** [comment]
It would be nice, too, if there was a way to fall back to `scope="session"` if pytest-xdist is disabled, without having to edit every fixture. I'm not sure what that would look like though. Maybe something silly like:

```
def node_scoped():
    try:
        import xdist
        return "node"
    except:
        return "session"
 
@fixture(scope=node_scoped())
def f(): ...
```

?

**Comment by RonnyPfannschmidt** [comment]
By now it's absolutely impossible to change the meaning of session

**Comment by kousu** [comment]
I've poked at this for the last 24 hours and I've got a couple of variants on a solution for what I want to call `scope="node"` -- making sure to only run once _per machine_. Prototypes of it, at least.

I haven't attempted to tackle the genuine `scope="session"` that this thread is about; potential cross-server, cross-OS testing makes that very complicated very fast. The use case I imagine for it is like, webapp testing, where you want to provision a single shared database for all the tests to try, but the tests themselves are distributed; you could put a lock in that database but making a generic solution is beyond me right now.

## tmpdir-based prototype

First, inspired by @kapilt's [work](https://github.com/cloud-custodian/pytest-terraform/blob/17bb7e8f87540333d65ccdd2554464c798036101/pytest_terraform/xdist.py) by there's this one that makes a lockfile inside of pytest's tmpdir. It should work with any fixture (I hope) and doesn't need any extra dependencies beyond pytest and the stdlib:

https://github.com/neuropoly/spinalcordtoolbox/blob/ng/pytest-nodescoped-via-tmpdir/conftest.py#L37-L162

Unfortunately, it's got some sharp metaprogramming edges because as far as I can tell I *must* use the `tmpdir` fixture in my fixture to find out where pytest wants me to work, and that meant messing around with `inspect.signature()` to make pytest's implicit fixture system happy. Oh and to support generator-fixtures, I also call the private `_pytest.fixtures.call_fixture_func()` which isn't ideal.

## uid-based protoype

The second makes a lockfile inside of the system's tempdir, but based on `testrun_uid`, which I am gratefully able to access with `os.environ`, meaning no messing around with fixture magics:

https://github.com/neuropoly/spinalcordtoolbox/blob/ng/pytest-nodescoped-via-uuid/conftest.py#L39-L153

I also figured out a better way to support generator fixtures in this one.

## some thoughts

Both of them pickle the result of the main worker for the blocked workers to pick up. I don't know if that's ideal, but it works.

You can see the latter version in action here: https://showterm.io/84e2112ea6b1955772a71#fast; the fixture is creating the `sct_testing_data/` folder, and the successful case is when it downloads the 7.17MB of it *once* with all the tests waiting for that to finish before beginning.

### some advice

Because of the auto-caching it's probably best to make sure your node-scoped fixtures return very small data. If your fixtures (as ours do) generate large binary files that you need to load, instead of doing

```
@fixture(scope="session")
# worker-scoped
def image(tmp_path):
    return make_lots_of_data()
```

which generates

or

```
@fixture(scope="session")
@node_scoped
def image(tmp_path):
    return make_lots_of_data()
```

which generates it once, but then tries to pickle the whole thing (does that even work with numpy arrays or Pillow images?)


You could chain them like this:

```
@fixture(scope="session")
@node_scoped
def disk_file(tmp_path):
    fname = tmp_path / "large_image.jpg.7z"
    with open(fname, "wb") as fd:
        make_lots_of_data().save(fd)
    return fname

@fixture(scope="session")
# *worker*-scoped
def image(disk_file):
    return Image.open(disk_file)
```

so only a small string is transmitted via pickle.

## even more thoughts

I **wanted** to use https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Queue but (from above) maybe that's still incompatible? https://github.com/pytest-dev/pytest-xdist/issues/271#issuecomment-429573187. Anyway, I couldn't get it to work. And it still needs pickle.

Maybe [mqueue](http://semanchuk.com/philip/posix_ipc/#message_queue) is what, but that would be an extra dependency, and it still needs pickle.

Maybe `os.mkfifo`? But unix pipes aren't reliable with multiple readers. Maybe I could make it work by making a separate one for each reader, but that seems hard to manage.

Maybe `mmap`? Or [`shm`](https://docs.python.org/3/library/multiprocessing.shared_memory.html#multiprocessing.shared_memory.ShareableList)?

**Comment by RonnyPfannschmidt** [comment]
the new scope names could be `test-run` and `test-node` to signify once per pytest invocation vs once per node running the tests

the session scope stays as is, as technically each worker runs a complete pytest session

as for logistics and communication, the result of such a fixture should be serializable either for json or for the backend (currently only execnet)

session scoped fixtures within a node code then take that data and turn it into real objects

**Comment by Yajo** [comment]
How about `xdist-session`?

**Comment by RonnyPfannschmidt** [comment]
Rejected, too close to session

**Comment by kousu** [comment]
`test-run`, `test-node` and `session` are good!

**Comment by kapilt** [comment]
@kousu afaics this seems to be missing some things I was doing in my impl for track fixture to test as dependency map to ensure tear down after last scoped usage of a fixture, which is why I ended up with a WAL on the test reporting protocol from xdist worker to controller process.  which makes me a little cautious about this implementation. afaics looking it over its racing in all worker processes to create (which is fine), *but* then its tearing down in any worker process that finishes tests, which may still result in multiple fixture initialization on a given node if an unfinished worker process still has to execute a test with the given fixture.

**Comment by leorochael** [comment]
What I've been using in lieu of locks is session fixtures is to define a `pytest_sessionstart(session)` on the root `conftest.py` of my project.

When running in parallel, this function will be called exactly once in the `master` process (which will not actually run any tests), then once in each xdist worker that is launched, before any session fixtures are run.

When not running in parallel, this function will be called only once.

The way to distinguish between the master is by the presence of a `workerinput` attribute in `session.config`, like this:

```python
import sys

def pytest_sessionstart(session):
    workerinput = getattr(session.config, 'workerinput', None)
    if workerinput is None:
        print("Running on Master Process, or not running in parallel at all.", file=sys.stderr)
    else:
        print(f"\nRunning on Worker: {workerinput['workerid']}\n", file=sys.stderr)
```

Besides access to  `workerinput`, the `session` parameter also gives access to the [pytest cache in `session.config.cache`](https://docs.pytest.org/en/stable/reference.html#config-cache)

Values set, and directories created, through the `session.config.cache` in the master process are available in workers, in particular in session fixtures, by requiring the `pytestconfig` fixture and accessing `pytestconfig.cache`.

Though the pytest cache use probably only works for `xdist` uses in the same machine (e.g by using the `-n` switch), rather than with more exotic uses of `execnet`, since it works by accessing the shared cache directory, unless `pytest-xdist` also syncs the `pytest` cache directory through `execnet` along with the tests themselves, which I haven't tested.

**Comment by leorochael** [comment]
By the way, there's also a `pytest_sessionend(session, exitstatus)`:

```
def pytest_sessionfinish(session, exitstatus):
    workerinput = getattr(session.config, 'workerinput', None)
    if workerinput is None:
        print("\nExiting the Master Process, or running serially", file=sys.__stderr__)
    else:
        print(f"\nExiting Worker: {workerinput['workerid']}\n", file=sys.__stderr__)
```

Notice the use of `sys.__stderr__` since `sys.stderr` is captured by this point.

This function is called once in each worker, strictly before it's being called on the master process.

**Comment by kousu** [comment]
@leorochael we tried that same workaround, but we found it was flakey. In rare cases, and more often on macOS, it runs multiple times: https://github.com/neuropoly/spinalcordtoolbox/issues/3071#issuecomment-752412772

I don't fully understand why. It seems like it should work. But that's why I went looking for other solutions.

@kapilt thanks for the feedback! I got sidetracked with solving other problems. I haven't looked into the bug you brought up with my implementation, but I believe you. It's hard to get concurrency stuff right.

**Comment by leorochael** [comment]
@kousu, if understand correctly, [your code does this](https://github.com/neuropoly/spinalcordtoolbox/blob/master/conftest.py#L28-L31):

```python
def pytest_sessionstart(session):
    """ Download sct_testing_data prior to test collection. """
    logger.info("Downloading sct test data")
    downloader.main(['-d', 'sct_testing_data', '-o', sct_test_path()])
```

But that code is **guaranteed** to run **multiple times**, once alone in the master process, then multiple times in parallel, once for each worker.

If you want to guarantee that code to run only once, in the master process only, you must do:

```python
def pytest_sessionstart(session):
    """ Download sct_testing_data prior to test collection. """
    if getattr(session.config, 'workerinput', None) is not None:
        # No need to download, the master process has already done that.
        return
    logger.info("Downloading sct test data")
    downloader.main(['-d', 'sct_testing_data', '-o', sct_test_path()])
```

Or if you want to save even more time between test runs on the same machine, check if the data isn't already downloaded before calling `downloader.main(...)`

```python
def pytest_sessionstart(session):
    """ Download sct_testing_data prior to test collection. """
    if getattr(session.config, 'workerinput', None) is not None:
        # No need to download, the master process has already done that.
        return
    test_path = sct_test_path()
    if os.path.isdir(test_path):
        logger.info("Already downloaded sct test data")
    logger.info("Downloading sct test data")
    downloader.main(['-d', 'sct_testing_data', '-o', test_path])
```

For other codebases that don't have their own infrastructure for locating where to save files, I recommend [using `session.config.cache.makedir()`](https://docs.pytest.org/en/stable/reference.html#pytest.Cache.makedir), then checking if the expected files are present inside the created directory, or using `session.config.get()` before calling `session.config.set()`

If you do that, you don't even need to check `getattr(session.config, 'workerinput', None)`, as the master process, which is guaranteed to run `pytest_sessionstart()` alone before launching workers, will have already downloaded the data before `pytest_sessionstart()` is launched in each worker in parallel.

I've made this flowchart to make the sequence of steps more clear:

[![PyTest XDist Session Flowchart](https://user-images.githubusercontent.com/239125/116010710-10a9ca00-a5f7-11eb-8e0e-22a417ba9dc8.png)](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiZ3JhcGggVERcbiAgICBBW1wiYHB5dGVzdF9zZXNzc2lvbnN0YXJ0KHNlc3Npb24pYDxiciAvPm9uIG1hc3Rlci48YnIgLz5gc2Vzc2lvbi5jb25maWcud29ya2VyaW5wdXRgIGF0dHJpYnV0ZSBpcyBtaXNzaW5nXCJdIC0tPlNwYXdue1NwYXduIHdvcmtlcnN9XG5cbiAgICBTcGF3biAtLT58Z3cwfCBHVzBfU3RhcnRbXCJgcHl0ZXN0X3Nlc3NzaW9uc3RhcnQoc2Vzc2lvbilgLjxiciAvPmBzZXNzaW9uLmNvbmZpZy53b3JrZXJpbnB1dGA8YnIvPmlzIGEgYGRpY3RgXCJdXG4gICAgU3Bhd24gLS0-fGd3MXwgR1cxX1N0YXJ0W1wiYHB5dGVzdF9zZXNzc2lvbnN0YXJ0KHNlc3Npb24pYC48YnIgLz5gc2Vzc2lvbi5jb25maWcud29ya2VyaW5wdXRgPGJyLz5pcyBhIGBkaWN0YFwiXVxuICAgIFNwYXduIC0tPnxndzJ8IEdXMl9TdGFydFtcImBweXRlc3Rfc2Vzc3Npb25zdGFydChzZXNzaW9uKWAuPGJyIC8-YHNlc3Npb24uY29uZmlnLndvcmtlcmlucHV0YDxici8-aXMgYSBgZGljdGBcIl1cblxuICAgIEdXMF9TdGFydCAtLT4gR1cwX1Nlc3Npb25fZml4dHVyZXNbXCJTZXNzaW9uIEZpeHR1cmVzXCJdXG4gICAgR1cwX1Nlc3Npb25fZml4dHVyZXMgLS0-IEdXMF9GaXh0dXJlc1xuc3ViZ3JhcGggR1cwX1Rlc3RfTG9vcCBbUnVuIFRlc3QgTG9vcF1cbiAgICBHVzBfRml4dHVyZXNbXCJPdGhlciBmaXh0dXJlc1wiXSAtLT4gR1cwX1Rlc3RzW1wiVGVzdHNcIl1cbmVuZFxuICAgIEdXMF9UZXN0cyAtLT4gR1cwX0VuZFtcImBweXRlc3Rfc2Vzc2lvbmZpbmlzaChzZXNzaW9uLCBleGl0c3RhdHVzKWAuPGJyIC8-YHNlc3Npb24uY29uZmlnLndvcmtlcmlucHV0YDxici8-aXMgYSBgZGljdGBcIl1cblxuICAgIEdXMV9TdGFydCAtLT4gR1cxX1Nlc3Npb25fZml4dHVyZXNbXCJTZXNzaW9uIEZpeHR1cmVzXCJdXG4gICAgR1cxX1Nlc3Npb25fZml4dHVyZXMgLS0-IEdXMV9GaXh0dXJlc1xuc3ViZ3JhcGggR1cxX1Rlc3RfTG9vcCBbUnVuIFRlc3RzXVxuICAgIEdXMV9GaXh0dXJlc1tcIk90aGVyIGZpeHR1cmVzXCJdIC0tPiBHVzFfVGVzdHNbXCJUZXN0c1wiXVxuZW5kXG4gICAgR1cxX1Rlc3RzIC0tPiBHVzFfRW5kW1wiYHB5dGVzdF9zZXNzaW9uZmluaXNoKHNlc3Npb24sIGV4aXRzdGF0dXMpYC48YnIgLz5gc2Vzc2lvbi5jb25maWcud29ya2VyaW5wdXRgPGJyLz5pcyBhIGBkaWN0YFwiXVxuXG4gICAgR1cyX1N0YXJ0IC0tPiBHVzJfU2Vzc2lvbl9maXh0dXJlc1tcIlNlc3Npb24gRml4dHVyZXNcIl1cbiAgICBHVzJfU2Vzc2lvbl9maXh0dXJlcyAtLT4gR1cyX0ZpeHR1cmVzXG5zdWJncmFwaCBHVzJfVGVzdF9Mb29wIFtSdW4gVGVzdHNdXG4gICAgR1cyX0ZpeHR1cmVzW1wiT3RoZXIgZml4dHVyZXNcIl0gLS0-IEdXMl9UZXN0c1tcIlRlc3RzXCJdXG5lbmRcbiAgICBHVzJfVGVzdHMgLS0-IEdXMl9FbmRbXCJgcHl0ZXN0X3Nlc3Npb25maW5pc2goc2Vzc2lvbiwgZXhpdHN0YXR1cylgLjxiciAvPmBzZXNzaW9uLmNvbmZpZy53b3JrZXJpbnB1dGA8YnIvPmlzIGEgYGRpY3RgXCJdXG5cbiAgICBHVzBfRW5kIC0tPiBKb2luXG4gICAgR1cxX0VuZCAtLT4gSm9pblxuICAgIEdXMl9FbmQgLS0-IEpvaW5cblxuICAgIEpvaW57XCJXYWl0IGZvciB3b3JrZXJzPGJyLz5Db2xsZWN0IHRlc3QgcmVzdWx0c1wifVxuICAgIEpvaW4gLS0-IEVuZFtcImBweXRlc3Rfc2Vzc2lvbmZpbmlzaChzZXNzaW9uLCBleGl0c3RhdHVzKWA8YnIgLz5vbiBtYXN0ZXIuPGJyIC8-YHNlc3Npb24uY29uZmlnLndvcmtlcmlucHV0YCBhdHRyaWJ1dGUgaXMgbWlzc2luZ1wiXSIsIm1lcm1haWQiOnt9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlfQ)

**Comment by leorochael** [comment]
By the way, when running in each worker, the content of the `session.config.workerinput` dictionary is something like:

```python
{
    'workerid': 'gw0',  # or gw1, gw2, etc...
    'workercount': 2,  # Or whatever you passed as argument to `pytest -n`
    'testrunuid': '2d1c(...)f350',
    'mainargv': ['/path/to/venv/bin/pytest', '-n', '2'],  # Or whatever your `pytest` cmdline was
}
```

So you can do things like create separated directory structures or restore dumps into separate databases, or whatever you need to do so that your tests running in different workers don't step on each other's toes by using `session.config.workerinput['workerid']` to create different names for your structures.

Please feel free to use any of the above to enhance the docs for `pytest-xdist`.

**Comment by ktbarrett** [comment]
Is there status on the implementation of the per-node and per-pytest-invocation scopes? I didn't seen anything linked in this thread, but perhaps something wasn't linked.

We are using pytest-xdist to run tests in parallel and want to share a single "build" step between those tests. The "session" scope doing a build per worker is _fine_, but it would be nice to instead do this only once per node to reduce disk usage and maybe a little time.

**Comment by nicoddemus** [comment]
Hi @ktbarrett, 

> Is there status on the implementation of the per-node and per-pytest-invocation scopes? I

AFAIK nobody is working on it actively.

**Comment by leorochael** [comment]
@ktbarrett, the way to run a setup only once per session, instead of once per worker is to declare a `pytest_sessionstart(session)` and check that `getattr(session.config, 'workerinput', None) is None`, as [I mentioned previously](https://github.com/pytest-dev/pytest-xdist/issues/271#issuecomment-826396320).

**Comment by ktbarrett** [comment]
@leorochael Thanks for the hint, but that's not a solution that I want to recommend to end users. Perhaps it's fine if I package that solution for them.

@nicoddemus It seems some support would need to be added to remote.py to run such fixtures, which doesn't seem that bad, the real problem seems to be registering a new scope with pytest. I didn't see that mentioned in the plugin documentation, do you have any idea? I am willing to work on this, but I would probably only work on the per-node scope.

**Comment by nicoddemus** [comment]
>  I didn't see that mentioned in the plugin documentation, do you have any idea? 

Currently it is not possible to add a new fixture scope via a plugin, I'm afraid.

I suggest instead you package your solution into a function/context manager of sorts, say a function which receives `config` and returns the fixture value, dealing with using a inter-process lock to ensure it is computed only once.

Say instead of users writing this:

```python
@pytest.fixture(scope="session")
def value():
    return compute_value_slow()
```

As is, each worker will execute its own copy of `compute_value_slow`.

From the top of my head, I guess something like this can be implemented:

```python
@pytest.fixture(scope="session")
def value():
    with ensure_unique_computation() as signal:
        if signal.compute:
            signal.value = compute_value_slow()
    return signal.value
```

`ensure_unique_computation` can then use an inter-process lock so:

1. The first process entering the context manager returns that `signal` object with `signal.compute == True` and `signal.value` as `None`, and acquires an inter-process lock. This tells the caller that they should compute the fixture value and assign it to the `signal.value` object.
2. The next processes entering the context manager, should wait on the inter-process lock.
3. The first process, when exiting the context manager, now has the computed value in `signal.value`. It should serialize it to disk somewhere (probably next to the lock), and release the lock.
4. The other processes, open getting the lock, should now be able to unserialize the object from disk, and assign them to their `signal.value` object.

Now that I think about it, it might even be something built into `pytest-xdist` itself, is somebody can make it work and provide a PR.

**Comment by ktbarrett** [comment]
Seeing as the body of `value` would be repeated besides the body of the `if signal.compute` branch, how about a decorator instead?

```python
@pytest.fixture(scope="session")
@xdist.per_node_fixture
def value():
    return compute_value_slow()
```

You could even merge the two decorators and provide a `xdist.fixture` that handles `scope="node"`. My gut tells me that the reaction to that would be "no".

> Now that I think about it, it might even be something built into pytest-xdist itself, if somebody can make it work and provide a PR.

Sure thing.

**Comment by RonnyPfannschmidt** [comment]
currently there is no safe way to have "per pytest combined testrun fixtures"

in particular since its actually moretricky, as you bascially have multiple dimensions of interactions

* hosts/nodes
* users
* just the normal parallelization

there are types of fixtures one would want to run once pe

right now im nto ware of a clearcut way to handle details on many of those

the common suggestions for now is to rely on parallelization and use locks/caches

while i'd love to see better coordination tools, i currently dont have a use-case for those myself, and not the spare cycles to attack this problem differently

**Comment by philippefutureboy** [comment]
I implemented a simple solution if one needs to run a side effect once using a file.
Not the best, but if that can help someone, here goes:

```py

@pytest.fixture(scope="session", autouse=True)
def before_all(tmp_path_factory):
    TMP_ROOT = tmp_path_factory.getbasetemp().parent
    LOCK_PATH = TMP_ROOT / 'before_all.lock'

    if os.environ.get('PYTEST_XDIST_WORKER') == 'gw0':
        shutil.rmtree("path/to/output/folder") # do your action
        open(LOCK_PATH, 'a').close()
    else:
        while not LOCK_PATH.is_file():
            time.sleep(0.1) # wait until the LOCK_PATH exists
```

If you need to clear up what you set up after your tests, you can write to the LOCK_PATH some sort of semaphore on exit, e.g. an integer. If the integer == the number of workers then you know you are in the last one before all the runs are done and you can then tear down the common infrastructure/state.
If you need to do this across a decentralized system, you could use a memcached or redis instance and save your semaphore as a key:value pair where the key is the key of the current run (e.g. current commit hash)

Hopefully that can help!

**Comment by nicoddemus** [comment]
Thanks for sharing @philippefutureboy,

There's a documented solution similar to yours but which is more robust in https://pytest-xdist.readthedocs.io/en/latest/how-to.html#making-session-scoped-fixtures-execute-only-once.

**Comment by ofek** [comment]
I hit this today. If it helps anyone, this is how I'm handling global setup/tear down of Docker containers https://github.com/pypa/hatch/pull/383/commits/0676efb80a5aef9338dbc8bc49249a27a22177b5

Essentially, I'm (de-)populating a global directory of worker IDs.

**Comment by maciejmatczak** [comment]
I was thinking about different solution.

Is it possible to group session-fixture-dependent tests? Somewhat automatically is preferred. Examples of my scenarios below.

I have a heavy to setup fixtures, that I later on reuse. I would like to see a parallelization (test grouping) to happen exactly around this fixtures, but reusage happening synchronously. I am spinning up a multiple servers, with which I need to communicate.

Simple example:

```python
import pytest
import time


@pytest.fixture(scope="session")
def heavy1():
    time.sleep(10)

    yield "heavy1"


@pytest.fixture(scope="session")
def heavy2():
    time.sleep(10)

    yield "heavy2"


def test_something(heavy1):
    assert heavy1 == "heavy"


def test_something_else(heavy1):
    assert heavy1 == "heavy1"


def test_different_stuff(heavy2):
    assert heavy2 == "heavy2"
```

I would like to see, when running with `-n2`, having 2 first tests in first worker and 3rd one in second worker. Because of that split, `heavy1` fixture would be setup once, also only in first worker. Analogically for `heavy2`.

That would parallelize the tests execution where I need it parallelized.

My real case scenario fixtures are more like:

```python
import pytest
import time

from foo import server


@pytest.fixture(scope="session", params=[
    "a",
    "b",
    "c",
])
def heavy(request):
    with server.start(param=request.param) as connection:
        yield connection

def test_something(heavy):
    assert heavy1 == "heavy"

def test_different_thing(heavy):
    assert heavy1 == "heavy1"
```

I parametrize fixture and I would like to group tests per effective fixture value... Like:

- `test_something[a]`, `test_different_thing[a]`
- `test_something[b`, `test_different_thing[b]`
- `test_something[c]`, `test_different_thing[c]`
